From 123c70cdb86e360c25519ca040a90e20ca648e8d Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Mon, 23 Jun 2014 21:51:23 -0400
Subject: [fs] nfsd: really fix nfs4err_resource in 4.1 case

Message-id: <1403560314-32525-142-git-send-email-bfields@redhat.com>
Patchwork-id: 84548
O-Subject: [Patch RHEL7 141/172] nfsd4: really fix nfs4err_resource in 4.1 case
Bugzilla: 1111719
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

encode_getattr, for example, can return nfserr_resource to indicate it
ran out of buffer space.  That's not a legal error in the 4.1 case.
And in the 4.1 case, if we ran out of buffer space, we should have
exceeded a session limit too.

(Note in 1bc49d83c37cfaf46be357757e592711e67f9809 "nfsd4: fix
nfs4err_resource in 4.1 case" we originally tried fixing this error
return before fixing the problem that we could error out while we still
had lots of available space.  The result was to trade one illegal error
for another in those cases.  We decided that was helpful, so reverted
the change in fc208d026be0c7d60db9118583fc62f6ca97743d, and are only
reinstating it now that we've elimited almost all of those cases.)

Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: c8f13d977518e588ac89dcf8e841821569108109
Bugzilla: 1111719
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4xdr.c b/fs/nfsd/nfs4xdr.c
index 480f12c..7f346d8 100644
--- a/fs/nfsd/nfs4xdr.c
+++ b/fs/nfsd/nfs4xdr.c
@@ -3901,6 +3901,14 @@ nfsd4_encode_operation(struct nfsd4_compoundres *resp, struct nfsd4_op *op)
    space_needed = COMPOUND_ERR_SLACK_SPACE;
   op->status = nfsd4_check_resp_size(resp, space_needed);
  }
+ if (op->status == nfserr_resource && nfsd4_has_session(&resp->cstate)) {
+  struct nfsd4_slot *slot = resp->cstate.slot;
+
+  if (slot->sl_flags & NFSD4_SLOT_CACHETHIS)
+   op->status = nfserr_rep_too_big_to_cache;
+  else
+   op->status = nfserr_rep_too_big;
+ }
  if (op->status == nfserr_resource ||
      op->status == nfserr_rep_too_big ||
      op->status == nfserr_rep_too_big_to_cache) {
-- 
1.7.1